mlflow/metrics/utils/metric_definitions.py

```diff
@@ -0,0 +1,409 @@
+import logging
+
+import pandas as pd
+
+from mlflow.metrics.base import MetricValue
+from mlflow.models import make_metric
+
+_logger = logging.getLogger(__name__)
+
+
+def _validate_text_predictions(predictions, metric_name):
+    if len(predictions) == 0:
+        return False
+
+    if any(not isinstance(prediction, str) for prediction in predictions):
+        _logger.warning(
+            f"Cannot calculate {metric_name} for non-string inputs, skipping metric logging."
+        )
+        return False
+
+    return True
+
+
+def _toxicity_eval_fn(eval_df, metrics):
+    y_pred = eval_df["prediction"]
+    predictions = y_pred.squeeze() if isinstance(y_pred, pd.DataFrame) else y_pred
+
+    if not _validate_text_predictions(predictions, "toxicity"):
+        return
+
+    try:
+        _logger.info("Loading toxicity metric:")
+        import evaluate
+
+        toxicity = evaluate.load("toxicity", module_type="measurement")
+    except Exception as e:
+        _logger.warning(
+            f"Failed to load 'toxicity' metric (error: {e!r}), skipping metric logging."
+        )
+        return
+
+    _logger.info("Computing toxicity metric:")
+    scores = toxicity.compute(predictions=predictions)["toxicity"]
+    toxicity_ratio = toxicity.compute(predictions=predictions, aggregation="ratio")[
+        "toxicity_ratio"
+    ]
+    return MetricValue(scores=scores, aggregate_results={"ratio": toxicity_ratio})
+
+
+def _perplexity_eval_fn(eval_df, metrics):
+    y_pred = eval_df["prediction"]
+    predictions = y_pred.squeeze() if isinstance(y_pred, pd.DataFrame) else y_pred
+
+    if not _validate_text_predictions(predictions, "perplexity"):
+        return
+
+    try:
+        _logger.info("Loading perplexity metric:")
+        import evaluate
+
+        perplexity = evaluate.load("perplexity", module_type="metric")
+    except Exception as e:
+        _logger.warning(
+            f"Failed to load 'perplexity' metric (error: {e!r}), skipping metric logging."
+        )
+        return
+
+    _logger.info("Computing perplexity metric:")
+    results = perplexity.compute(predictions=predictions, model_id="gpt2")
+    return MetricValue(
+        scores=results["perplexities"], aggregate_results={"mean": results["mean_perplexity"]}
+    )
```

Comment:
for a given set of `scores`, don't we know all the various aggregations (min, max, mean, median, variance, p90)? if so, does it make sense to compute and store all of the aggregations in the resulting `MetricValue`

this comment applies to the other metric eval_fns in this file too

Reply:
yeah right now i'm moving over all the current functionality from mlflow.evaluate() to use make_metric() without adding/changing anything that is calculated - planning to have a follow-up PR to calculate/store those aggregations

Reply:
ah understood. that makes sense.